File: frontend/src/api/studentImport.ts
import { apiClient, downloadFile } from './client'

export type ImportMode = 'delete_all' | 'update_or_add' | 'new_only'

export interface AnalyzeResponse {
  columns: string[]
  suggested_mapping: Record<string, string | null>
  available_fields: Record<string, string>
  raw_rows: Record<string, unknown>[]
  total_rows: number
  distinct_class_values: string[]
  suggested_class_mapping: Record<string, string | null>
  known_classes: string[]
}

export interface PreviewRow {
  row_number: number
  data: Record<string, unknown>
  status: 'valid' | 'invalid' | 'duplicate'
  errors: string[]
  missing_fields: string[]
}

export interface PreviewResponse {
  total_rows: number
  valid_rows: number
  invalid_rows: number
  duplicate_rows: number
  missing_fields_rows: number
  rows: PreviewRow[]
}

export interface ExecuteResult {
  imported: number
  updated: number
  skipped: number
  failed: number
  errors: { row_number: number; reason: string }[]
}

export async function analyzeImportFile(file: File): Promise<AnalyzeResponse> {
  const formData = new FormData()
  formData.append('file', file)
  const { data } = await apiClient.post<AnalyzeResponse>('/students/import/analyze', formData)
  return data
}

export async function previewImport(payload: {
  raw_rows: Record<string, unknown>[]
  mapping: Record<string, string | null>
  class_value_mapping: Record<string, string>
  import_mode: ImportMode
}): Promise<PreviewResponse> {
  const { data } = await apiClient.post<PreviewResponse>('/students/import/preview', payload)
  return data
}

export async function executeImport(payload: {
  raw_rows: Record<string, unknown>[]
  mapping: Record<string, string | null>
  class_value_mapping: Record<string, string>
  import_mode: ImportMode
  only_valid_rows: boolean
  confirm_delete_all: boolean
}): Promise<ExecuteResult> {
  const { data } = await apiClient.post<ExecuteResult>('/students/import/execute', payload)
  return data
}

export type ExportScope = 'all' | 'selected' | 'filtered' | 'search'

export async function exportStudents(params: {
  format: 'xlsx' | 'csv'
  scope: ExportScope
  student_ids?: number[]
  search?: string
  class_filter?: string
  status_filter?: string
}): Promise<void> {
  await downloadFile('/students/export', {
    format: params.format,
    scope: params.scope,
    student_ids: params.student_ids?.join(',') ?? '',
    search: params.search ?? '',
    class_filter: params.class_filter ?? '',
    status_filter: params.status_filter ?? '',
  }, `students_export.${params.format}`)
}

File: backend/app/schemas/student_import.py
from typing import Any, Literal

from pydantic import BaseModel

ImportMode = Literal["delete_all", "update_or_add", "new_only"]

IMPORT_FIELDS: dict[str, str] = {
    "registration_no": "Registration Number",
    "name": "Student Name",
    "father_name": "Father's Name",
    "class_name": "Class",
    "phone": "Parent Phone",
    "cnic": "CNIC",
    "default_fee": "Monthly Fee",
    "dob": "Date of Birth",
    "admission_date": "Admission Date",
    "b_form": "B-Form Number",
    "address": "Address",
    "status": "Status",
}


class AnalyzeResponse(BaseModel):
    columns: list[str]
    suggested_mapping: dict[str, str | None]
    available_fields: dict[str, str] = IMPORT_FIELDS
    raw_rows: list[dict[str, Any]]
    total_rows: int
    distinct_class_values: list[str]
    suggested_class_mapping: dict[str, str | None]
    known_classes: list[str]


class PreviewRequest(BaseModel):
    raw_rows: list[dict[str, Any]]
    mapping: dict[str, str | None]  # excel column -> our field key (None = ignore this column)
    class_value_mapping: dict[str, str]  # raw class value -> canonical class_name
    import_mode: ImportMode


class PreviewRow(BaseModel):
    row_number: int
    data: dict[str, Any]
    status: Literal["valid", "invalid", "duplicate"]
    errors: list[str] = []
    missing_fields: list[str] = []


class PreviewResponse(BaseModel):
    total_rows: int
    valid_rows: int
    invalid_rows: int
    duplicate_rows: int
    missing_fields_rows: int
    rows: list[PreviewRow]


class ExecuteRequest(BaseModel):
    raw_rows: list[dict[str, Any]]
    mapping: dict[str, str | None]
    class_value_mapping: dict[str, str]
    import_mode: ImportMode
    only_valid_rows: bool = True
    confirm_delete_all: bool = False


class ExecuteError(BaseModel):
    row_number: int
    reason: str


class ExecuteResult(BaseModel):
    imported: int
    updated: int
    skipped: int
    failed: int
    errors: list[ExecuteError]
